aerleon/lib/proxmox.py

```diff
@@ -0,0 +1,658 @@
+
+        # proxmox firewall only supports one protocol per rule
+        for protocol in term_protocol:
+            # proxmox firewall does not support mixed AFs in filters
+            for af in address_families:
+                for icmp_type in icmp_types:
+                    for icmp_code in icmp_codes:
+                        source = self._ComputeAddresses(
+                            self.filter_for_af(af, self.term.source_address),
+                            self.term.source_address_exclude,
+                        )
+                        dest = self._ComputeAddresses(
+                            self.filter_for_af(af, self.term.destination_address),
+                            self.term.destination_address_exclude,
+                        )
+                        ret_str.append(
+                            self._Format(
+                                protocol,
+                                self.direction,
+                                self.ACTIONS_MAP[self.term.action[0]],
+                                source,
+                                dest,
+                                icmp_code,
+                                icmp_type,
+                                self.term.source_interface,
+                                self.term.source_port,
+                                self.term.destination_port,
+                                self.term.comment,
+                                self.term.logging,
+                                self.term.option,
+                            )
+                        )
+
+        return '\n'.join(ret_str)
+
+    def _Format(
+        self,
+        protocol: Union[str, None],
+        direction: str,
+        action: str,
+        source_addresses: List[Union[IPv4, IPv6]],
+        destination_addresses: List[Union[IPv4, IPv6]],
+        icmp_code: Union[str, None],
+        icmp_type: Union[str, None],
+        source_interface: str,
+        source_ports: List[Union[Tuple[int, int], int]],
+        destination_ports: List[Union[Tuple[int, int], int]],
+        comment: List[str],
+        logging: List[str],
+        term_options: List[str],
+    ):
+        def to_network_addr(i: Union[IPv6, IPv4]):
+            return str(i.with_prefixlen)
+
+        options = [direction, action]
+
+        if protocol:
+            options.append("-proto %s" % self.PROXMOX_PROTO_MAP.get(protocol, protocol))
+
+        # proxmox firewall supports multiple sources/destinations per rule
+        if source_interface:
+            options.append("-iface %s" % source_interface)
+
+        # we cannot use address tokens since IPsets are defined in separate files (at cluster
+        # or at host level) and we're only outputting one file at a time, which may not be
+        # cluster/host firewall configuration.
+        if destination_addresses:
+            options.append("-dest %s" % ','.join(map(to_network_addr, destination_addresses)))
```

Comment:
prefer f strings here, they are generally faster and lead to more readable code.